bitcoin/crawler.go

```diff
@@ -0,0 +1,539 @@
+		properties["connect_error"] = bitcoinResult.ConnectError.Error()
+	}
+
+	// keep track of all unknown crawl errors
+	if bitcoinResult.ErrorStr == models.NetErrorUnknown && bitcoinResult.Error != nil {
+		properties["crawl_error"] = bitcoinResult.Error.Error()
+	}
+
+	data, err := json.Marshal(properties)
+	if err != nil {
+		log.WithError(err).WithField("properties", properties).Warnln("Could not marshal peer properties")
+	}
+
+	if len(bitcoinResult.ListenAddrs) > 0 {
+		task.AddrInfo.Addr = bitcoinResult.ListenAddrs
+	}
+
+	cr := core.CrawlResult[PeerInfo]{
+		CrawlerID:           c.id,
+		Info:                task,
+		CrawlStartTime:      crawlStart,
+		RoutingTableFromAPI: false,
+		RoutingTable:        bitcoinResult.RoutingTable,
+		Agent:               bitcoinResult.Agent,
+		Protocols:           bitcoinResult.Protocols,
+		ConnectError:        bitcoinResult.ConnectError,
+		ConnectErrorStr:     bitcoinResult.ConnectErrorStr,
+		CrawlError:          bitcoinResult.Error,
+		CrawlErrorStr:       bitcoinResult.ErrorStr,
+		CrawlEndTime:        time.Now(),
+		ConnectStartTime:    bitcoinResult.ConnectStartTime,
+		ConnectEndTime:      bitcoinResult.ConnectEndTime,
+		Properties:          data,
+		LogErrors:           c.cfg.LogErrors,
+	}
+
+	// We've now crawled this peer, so increment
+	c.crawledPeers++
+
+	return cr, nil
+}
+
+func (c *Crawler) PeerProperties(node *AddrInfo) map[string]any {
+	// TODO: to be implemented later
+	properties := map[string]any{}
+
+	properties["Services"] = "TBI"
+
+	return properties
+}
+
+type BitcoinResult struct {
+	ConnectStartTime time.Time
+	ConnectEndTime   time.Time
+	ConnectError     error
+	ConnectErrorStr  string
+	Agent            string
+	ProtocolVersion  int32
+	Protocols        []string
+	ListenAddrs      []ma.Multiaddr
+	Error            error
+	ErrorStr         string
+	RoutingTable     *core.RoutingTable[PeerInfo]
+}
+
+func (c *Crawler) crawlBitcoin(ctx context.Context, pi PeerInfo) chan BitcoinResult {
+	resultCh := make(chan BitcoinResult)
+
+	go func() {
+		result := BitcoinResult{}
+		neighbours := make([]PeerInfo, 0)
+
+		addrs := pi.Addrs()
+
+		// addrInfo := peer.AddrInfo{
+		// 	ID:    pi.ID(),
+		// 	Addrs: sanitizedAddrs,
+		// }
+
+		var conn net.Conn
+		var err error
+		connectionMaxRetry := 10
+		result.ConnectStartTime = time.Now()
+		if len(addrs) > 0 {
+			for i := 0; i < connectionMaxRetry; i++ {
+				netAddr, _ := manet.ToNetAddr(addrs[0])
+				d := net.Dialer{
+					Timeout: c.cfg.DialTimeout,
+				}
+				conn, err = d.DialContext(ctx, netAddr.Network(), netAddr.String())
+				if err == nil {
+					break
+				}
+			}
+		}
+
+		if conn == nil {
+			result.RoutingTable = &core.RoutingTable[PeerInfo]{
+				PeerID:    pi.ID(),
+				Neighbors: neighbours,
+				ErrorBits: uint16(0), // FIXME
+				Error:     result.Error,
+			}
+			result.Error = err
+			resultCh <- result
+			close(resultCh)
```

Comment:
rather do a `defer close(resultCh)` and also remove the `close` call at the end of the function